File: server/ssrf.test.ts
import assert from 'node:assert/strict';
import dns from 'node:dns/promises';
import { afterEach, mock, test } from 'node:test';

import { assertSafeUrl } from './ssrf.ts';

afterEach(() => {
  mock.restoreAll();
});

// Helper: assert the promise rejects with a message matching `msg`.
async function rejects(p: Promise<unknown>, msg: string | RegExp) {
  await assert.rejects(p, (err: Error) => {
    assert.match(err.message, typeof msg === 'string' ? new RegExp(msg) : msg);
    return true;
  });
}

test('rejects malformed URLs', async () => {
  await rejects(assertSafeUrl('not a url'), 'Invalid URL');
  await rejects(assertSafeUrl(''), 'Invalid URL');
});

test('rejects non-http(s) protocols', async () => {
  await rejects(assertSafeUrl('ftp://example.com/x'), 'Only http/https');
  await rejects(assertSafeUrl('file:///etc/passwd'), 'Only http/https');
  await rejects(assertSafeUrl('gopher://example.com'), 'Only http/https');
});

test('blocks literal private/loopback/link-local IPv4 addresses', async () => {
  for (const ip of [
    '127.0.0.1',
    '10.1.2.3',
    '192.168.0.1',
    '172.16.5.5',
    '169.254.169.254', // cloud metadata endpoint
    '100.64.0.1', // CGNAT
    '0.0.0.0',
  ]) {
    await rejects(assertSafeUrl(`http://${ip}/`), 'Blocked address');
  }
});

test('blocks literal private/loopback IPv6 addresses', async () => {
  for (const host of [
    '[::1]', // loopback
    '[::]', // unspecified
    '[fc00::1]', // unique-local
    '[fe80::1]', // link-local
    // IPv4-mapped literals: the URL parser rewrites these to hex-compressed form
    // (e.g. [::ffff:7f00:1]), which must still be decoded back to the private IPv4.
    '[::ffff:127.0.0.1]', // → loopback
    '[::ffff:10.0.0.1]', // → private
    '[::ffff:169.254.169.254]', // → cloud metadata
  ]) {
    await rejects(assertSafeUrl(`http://${host}/`), 'Blocked address');
  }
});

test('blocks IPv4-mapped IPv6 addresses returned by DNS (dotted form)', async () => {
  // dns.lookup can yield the dotted `::ffff:a.b.c.d` form, which the guard
  // strips back to IPv4 before classifying.
  for (const address of ['::ffff:127.0.0.1', '::ffff:10.0.0.1']) {
    mock.method(dns, 'lookup', async () => [{ address, family: 6 }]);
    await rejects(assertSafeUrl('http://mapped.example.com/'), 'Blocked address');
    mock.restoreAll();
  }
});

test('still allows public IPv4-mapped IPv6 addresses', async () => {
  // The mapped-address path must not over-block: a mapped *public* IPv4 is fine.
  await assertSafeUrl('http://[::ffff:8.8.8.8]/');
});

test('allows literal public IP addresses without DNS lookup', async () => {
  const lookup = mock.method(dns, 'lookup');
  await assertSafeUrl('http://8.8.8.8/');
  await assertSafeUrl('https://1.1.1.1/path');
  await assertSafeUrl('http://[2606:4700:4700::1111]/'); // public IPv6
  assert.equal(lookup.mock.callCount(), 0, 'literal IPs skip DNS resolution');
});

test('allows hostnames that resolve to public addresses', async () => {
  mock.method(dns, 'lookup', async () => [{ address: '93.184.216.34', family: 4 }]);
  await assertSafeUrl('https://example.com/page');
});

test('blocks hostnames that resolve to a private address', async () => {
  // DNS-rebinding-style: a public-looking host pointing at the internal network.
  mock.method(dns, 'lookup', async () => [{ address: '10.0.0.5', family: 4 }]);
  await rejects(assertSafeUrl('http://evil.example.com/'), 'Blocked address');
});

test('blocks when any resolved address is private', async () => {
  mock.method(dns, 'lookup', async () => [
    { address: '93.184.216.34', family: 4 },
    { address: '192.168.1.1', family: 4 },
  ]);
  await rejects(assertSafeUrl('http://mixed.example.com/'), 'Blocked address');
});

test('rejects hostnames that do not resolve', async () => {
  mock.method(dns, 'lookup', async () => []);
  await rejects(assertSafeUrl('http://nxdomain.invalid/'), 'did not resolve');
});

File: server-go/internal/ssrf/ssrf_test.go
package ssrf_test

import (
	"context"
	"net"
	"strings"
	"testing"

	"rss-reader/server-go/internal/ssrf"
)

func assertErr(t *testing.T, err error, want string) {
	t.Helper()
	if err == nil {
		t.Fatalf("expected error containing %q, got nil", want)
	}
	if !strings.Contains(err.Error(), want) {
		t.Fatalf("error %q does not contain %q", err.Error(), want)
	}
}

func assertOK(t *testing.T, err error) {
	t.Helper()
	if err != nil {
		t.Fatalf("expected no error, got %v", err)
	}
}

func ctx() context.Context { return context.Background() }

func TestRejectsMalformedURLs(t *testing.T) {
	assertErr(t, ssrf.AssertSafeURL(ctx(), "not a url"), "Invalid URL")
	assertErr(t, ssrf.AssertSafeURL(ctx(), ""), "Invalid URL")
}

func TestRejectsNonHTTPProtocols(t *testing.T) {
	assertErr(t, ssrf.AssertSafeURL(ctx(), "ftp://example.com/x"), "Only http/https")
	assertErr(t, ssrf.AssertSafeURL(ctx(), "file:///etc/passwd"), "Only http/https")
	assertErr(t, ssrf.AssertSafeURL(ctx(), "gopher://example.com"), "Only http/https")
}

func TestBlocksLiteralPrivateIPv4(t *testing.T) {
	for _, ip := range []string{
		"127.0.0.1", "10.1.2.3", "192.168.0.1", "172.16.5.5",
		"169.254.169.254", // cloud metadata
		"100.64.0.1",      // CGNAT
		"0.0.0.0",
	} {
		assertErr(t, ssrf.AssertSafeURL(ctx(), "http://"+ip+"/"), "Blocked address")
	}
}

func TestBlocksLiteralPrivateIPv6(t *testing.T) {
	for _, host := range []string{
		"[::1]", "[::]", "[fc00::1]", "[fe80::1]",
		"[::ffff:127.0.0.1]",       // → loopback
		"[::ffff:10.0.0.1]",        // → private
		"[::ffff:169.254.169.254]", // → cloud metadata
	} {
		assertErr(t, ssrf.AssertSafeURL(ctx(), "http://"+host+"/"), "Blocked address")
	}
}

func TestBlocksMappedIPv4FromDNS(t *testing.T) {
	for _, addr := range []string{"::ffff:127.0.0.1", "::ffff:10.0.0.1"} {
		restore := ssrf.SetResolver(func(context.Context, string) ([]net.IP, error) {
			return []net.IP{net.ParseIP(addr)}, nil
		})
		assertErr(t, ssrf.AssertSafeURL(ctx(), "http://mapped.example.com/"), "Blocked address")
		restore()
	}
}

func TestAllowsPublicMappedIPv6(t *testing.T) {
	assertOK(t, ssrf.AssertSafeURL(ctx(), "http://[::ffff:8.8.8.8]/"))
}

func TestAllowsLiteralPublicIPsWithoutDNS(t *testing.T) {
	called := false
	restore := ssrf.SetResolver(func(context.Context, string) ([]net.IP, error) {
		called = true
		return nil, nil
	})
	defer restore()
	assertOK(t, ssrf.AssertSafeURL(ctx(), "http://8.8.8.8/"))
	assertOK(t, ssrf.AssertSafeURL(ctx(), "https://1.1.1.1/path"))
	assertOK(t, ssrf.AssertSafeURL(ctx(), "http://[2606:4700:4700::1111]/")) // public IPv6
	if called {
		t.Fatal("literal IPs must skip DNS resolution")
	}
}

func TestAllowsPublicResolvingHost(t *testing.T) {
	restore := ssrf.SetResolver(func(context.Context, string) ([]net.IP, error) {
		return []net.IP{net.ParseIP("93.184.216.34")}, nil
	})
	defer restore()
	assertOK(t, ssrf.AssertSafeURL(ctx(), "https://example.com/page"))
}

func TestBlocksHostResolvingToPrivate(t *testing.T) {
	restore := ssrf.SetResolver(func(context.Context, string) ([]net.IP, error) {
		return []net.IP{net.ParseIP("10.0.0.5")}, nil
	})
	defer restore()
	assertErr(t, ssrf.AssertSafeURL(ctx(), "http://evil.example.com/"), "Blocked address")
}

func TestBlocksWhenAnyResolvedIsPrivate(t *testing.T) {
	restore := ssrf.SetResolver(func(context.Context, string) ([]net.IP, error) {
		return []net.IP{net.ParseIP("93.184.216.34"), net.ParseIP("192.168.1.1")}, nil
	})
	defer restore()
	assertErr(t, ssrf.AssertSafeURL(ctx(), "http://mixed.example.com/"), "Blocked address")
}

func TestRejectsNonResolvingHost(t *testing.T) {
	restore := ssrf.SetResolver(func(context.Context, string) ([]net.IP, error) {
		return []net.IP{}, nil
	})
	defer restore()
	assertErr(t, ssrf.AssertSafeURL(ctx(), "http://nxdomain.invalid/"), "did not resolve")
}
